### ISSUE TITLE
Add a rule to enforce array `some`/`every` predicates return boolean

### ISSUE THREAD

**Issue by danielnixon** [typescript-eslint/typescript-eslint#2428]
**Repro**

```ts
const shouldWeDoIt = () => true;

// Whoops I forgot to include parens to call the function.
// Luckily strict-boolean-expressions catches it.
if (shouldWeDoIt) {
    
}

// But sadly these aren't caught, because the type of these predicates is a function that returns `unknown`, NOT `boolean`.
// I understand why the TypeScript team went this way, but ideally these predicates would require a `boolean` return type.
[].every(() => shouldWeDoIt);
[].some(() => shouldWeDoIt);
```

**Expected Result**

These predicate cases are caught by `strict-boolean-expressions` (or if this is considered outside the scope of this rule, by some other new rule).

**Actual Result**

Surprising runtime behavior.

**Comment by bradzacher** [comment]
This is well outside the scope of `strict-boolean-expressions`.
Though I wouldn't say no to a new rule which enforces strict array predicate returns.

**Comment by tadhgmister** [comment]
see https://github.com/microsoft/TypeScript/issues/19456 for some discussion about this within typescript itself.

**Comment by bradzacher** [comment]
Just realised that this is a duplicate of #1038

**Comment by danielnixon** [comment]
Even with this rule enabled:
```ts
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
```

Things like this are _not_ flagged:
```ts
[0].some((a) => a);
```


eslint 7.8.1, typescript-eslint 4.0.1, typescript 4.0.2

**Comment by danielnixon** [comment]
So I suspect we're still missing a rule (or an option to a rule) that combines `strict-boolean-expressions` with `no-unnecessary-condition` which is what I had in mind originally ("but ideally these predicates would require a `boolean` return type").

**Comment by bradzacher** [comment]
`[0]` is inferred to be `number[]`, so `a => a` is inferred to be `number => number`. `number` is both possibly falsy and possibly truthy, so the rule does not report.

OTOH if the array was typed as a tuple `[0]`, then the rule would report as `0` is only ever falsy.

---

> So I suspect we're still missing a rule

I take back what I said earlier - this _could_ live within `strict-boolean-expressions`, just like there is a check in `no-unnecessary-condition`. This is covered by #1038

**Comment by danielnixon** [comment]
Thanks @bradzacher !

**Comment by tadhgmister** [comment]
> This is covered by #1038

only partially, the `no-unnecessary-condition` checks that there is always truthy or falsy branches in conditions, strict boolean check is stricter than that. I think reusing some of the logic implemented in #1206 to ensure predicate functions always return a boolean (or alternate versions with options) is reasonable request.
